src/ast/irbuilderbpf.cpp

```diff
@@ -473,15 +473,20 @@ CallInst *IRBuilderBPF::CreateGetStrScratchMap(int idx,
                              idx);
 }
 
-CallInst *IRBuilderBPF::CreateGetFmtStringArgsScratchMap(
-    BasicBlock *failure_callback,
-    const location &loc)
-{
-  return createGetScratchMap(to_string(MapType::FmtStringArgs),
-                             "fmtstr",
-                             GET_PTR_TY(),
-                             loc,
-                             failure_callback);
+Value *IRBuilderBPF::CreateGetFmtStringArgsScratchBuffer(const location &loc)
+{
+  // Get CPU ID
+  auto cpu_id = CreateGetCpuId(loc);
```

Comment:
Just calling out that helper call here is fine. Verifier will inline the call so at runtime function call overhead is avoided.

Also, when we add the masking (to prove to older verifiers we are inbounds), we might wanna consider loading value into a register and doing the masking once at the beginning of the probe. That way we avoid re-doing the same work over and over if there's a lot of scratch map lookups.

Reply:
Will try that!

Masking didn't work as array size isn't always guaranteed to be a power of 2 - 1. I did what was suggested in the wiki and this worked - tested with 1000 printfs:
```
  auto max = getInt64(bpftrace_.resources.max_fmtstring_args_size * (bpftrace_.ncpus_ - 1));
  // check whether expr_ is less-than-or-equal-to maximum
  auto cmp = CreateICmp(CmpInst::ICMP_ULE, mult, max, "globalvar.min.cmp");
  // Select will contain expr_ if expr_ is sufficiently low, otherwise it will contain max
  auto select = CreateSelect(cmp, mult, max, "globalvar.min.select");
```

Reply:
Hmmm so I tried two things:
1. Bounding CPU ID between 0 and maxCpuId, then multiplying by max_fmt_args_size
2. Multiplying CPU ID by max_fmt_arg_size, then applying bound between 0 and maxCpuId*max_fmt_args_size

#1 didn't work on 5.2, verifier still barfed. #2 did.

Problem is for #2, we'll need to store a value per scratch map as each array is different sizes. With 11 registers and ~5 scratch maps, that takes up a good amount of registers.

For now, I'll leave it un-cached but happy to discuss further.

Reply:
I was unable to get caching into registers to work, I kept hitting exceptions when loading the BPF program. For now, I'm not gonna cache it, but happy to discuss offline if I'm missing something.